File: programmers/c++/단어변환.cpp
//단어변환
//dfs로 품
//37분

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool check[51];
int diff = 0;
int minValue = 987987987;
void dfs(int cnt, string word, vector<string> words, string target) {
	if (word == target) {
		minValue = min(minValue, cnt);
		return;
	}
	for (int j = 0; j < words.size(); j++) {
		if (check[j])
			continue;
		diff = 0;
		for (int i = 0; i < words[j].length(); i++) {
			if (word[i] != words[j][i])
				diff++;
		}
		if (diff == 1) {
			check[j] = true;
			dfs(cnt + 1, words[j], words, target);
			check[j] = false;
		}
	}
	return;
}
int solution(string begin, string target, vector<string> words) {
	int answer = 0;
	dfs(0, begin, words, target);
	if (minValue == 987987987)
		answer = 0;
	else
		answer = minValue;
	return answer;
}

File: programmers/고득점_kit_python/bfs_dfs/단어변환.py
#첫번째 풀이
visited = [0]*51
min_value = 987987987

def dfs(result, begin, target, words):
    global min_value 
    if begin == target:
        min_value = min(min_value, result)
        return
    for i, word in enumerate(words):
        cnt = 0
        for j, c in enumerate(word):
            if begin[j] != c:
                cnt += 1
        if cnt == 1 and not visited[i]:
            visited[i] = 1
            dfs(result + 1, word, target, words)
            visited[i] = 0
            
def solution(begin, target, words):
    global min_value 
    dfs(0, begin, target, words)
    if min_value == 987987987:
        return 0
    return min_value
